tick/preprocessing/longitudinal_features_product.py

```diff
@@ -0,0 +1,236 @@
+    --------
+    >>> from scipy.sparse import csr_matrix
+    >>> from tick.preprocessing.longitudinal_features_product import LongitudinalFeaturesProduct
+    >>> infinite_exposures = [csr_matrix([[0, 1, 0],
+    ...                                   [0, 0, 0],
+    ...                                   [0, 0, 1]], dtype="float64"),
+    ...                       csr_matrix([[1, 1, 0],
+    ...                                   [0, 0, 1],
+    ...                                   [0, 0, 0]], dtype="float64")
+    ...                       ]
+    >>> lfp = LongitudinalFeaturesProduct(exposure_type="infinite")
+    >>> product_features = lfp.fit_transform(features)
+    >>> # output comes as a list of sparse matrices or 2D numpy arrays
+    >>> product_features.__class__
+    <class 'scipy.sparse.csr.csr_matrix'>
+    >>> [x.toarray() for x in product_features] 
+    [array([[ 0.,  1.,  0.,  0.,  0.,  0.],
+            [ 0.,  0.,  0.,  0.,  0.,  0.],
+            [ 0.,  0.,  1.,  0.,  0.,  1.]]),
+    array([[ 1.,  1.,  0.,  1.,  0.,  0.],
+           [ 0.,  0.,  1.,  0.,  1.,  1.],
+           [ 0.,  0.,  0.,  0.,  0.,  0.]])]
+    """
+
+    _attrinfos = {
+        "exposure_type": {"writable": False},
+        "_mapper": {"writable": False},
+        "_n_init_features": {"writable": False},
+        "_n_output_features": {"writable": False},
+        "_n_intervals": {"writable": False},
+        "_preprocessor": {"writable": False},
+        "_fitted": {"writable": False}
+    }
+
+    def __init__(self, exposure_type="infinite", n_jobs=-1):
+        Base.__init__(self)
+        if exposure_type not in ["infinite", "short"]:
+            raise ValueError("exposure_type should be either 'infinite' or\
+             'short', not %s" %exposure_type)
+        self.n_jobs = n_jobs
+        self.exposure_type = exposure_type
+        self.reset()
+
+    def reset(self):
+        self._set("_n_init_features", None)
+        self._set("_n_output_features", None)
+        self._set("_n_intervals", None)
+        self._set("_mapper", {})
+        self._set("_preprocessor", None)
+        self._set("_fitted", False)
+
+    @property
+    def mapper(self):
+        """Get the mapping between the feature products and column indexes.
+
+        Returns
+        -------
+        output : `dict`
+            The column index - feature mapping.
+        """
+        if not self._fitted:
+            raise ValueError("cannot get mapper if object has not been fitted.")
+        return self._mapper.copy()
+
+    def fit(self, X):
+        """Fit the feature product using the features matrices list.
+
+        Parameters
+        ----------
+        X : `[np.ndarrays]`  or `[csr_matrices]`, shape=(n_intervals, n_features)
+            The list of features matrices.
+
+        Returns
+        -------
+        output : `LongitudinalFeaturesProduct`
+            The fitted current instance.
+        """
+        self.reset()
+        # TODO: check input type ?
```

Comment:
Definitively. You need to check that all shapes are equal and use _safe_array from our inference tools